funsor/gaussian.py

```diff
@@ -982,50 +958,150 @@ def _sample(self, sampled_vars, sample_inputs, rng_key):
+                ]
+            )
+            white_vec_a = (
+                Tensor(self.white_vec, int_inputs)
+                - (flat[None] @ Tensor(prec_sqrt_b, int_inputs))[0]
+            )
+
+            # Triangularize.
+            precision_chol_a = Tensor(ops.cholesky(_mmt(prec_sqrt_a)), int_inputs)
+            white_vec_a = ops.triangular_solve(
+                Tensor(prec_sqrt_a, int_inputs) @ white_vec_a[..., None],
+                precision_chol_a,
+            )[..., 0]
+
+            # Jointly sample.
+            white_noise = _sample_white_noise(
+                sample_inputs, int_inputs, dim_a, self.white_vec, rng_key
+            )
+            if not isinstance(white_noise, Funsor):
+                inputs = sample_inputs.copy()
+                inputs.update(int_inputs)
+                white_noise = Tensor(white_noise, inputs)
+            sample = ops.triangular_solve(
+                (white_noise + white_vec_a)[..., None], precision_chol_a, transpose=True
+            )[..., 0]
+
+            # Compute the remaining Gaussian, equivalent to
+            # self.reduce(ops.logaddexp, sampled_vars), but avoiding duplicate work.
+            inputs = int_inputs.copy()
+            inputs.update(remaining_real_inputs)
+            remaining = self._marginalize_after_split(
+                inputs, int_inputs, prec_sqrt_a, prec_sqrt_b, precision_chol_a.data
+            )
+
+        # Extract shaped components of the flat concatenated sample.
+        results = [remaining]
+        offsets, _ = _compute_offsets(sampled_real_inputs)
+        for key, domain in sampled_real_inputs.items():
+            point = sample[..., offsets[key] : offsets[key] + domain.num_elements]
+            point = point.reshape(point.shape[:-1] + domain.shape)
+            if not isinstance(point, Funsor):  # I.e. when eagerly sampling.
+                inputs = sample_inputs.copy()
+                inputs.update(int_inputs)
+                point = Tensor(point, inputs)
+            assert point.output == domain
+            results.append(Delta(key, point))
+
+        return reduce(ops.add, results)
+
+    def _marginalize_after_split(
+        self, inputs, int_inputs, prec_sqrt_a, prec_sqrt_b, precision_chol_a
+    ):
+        """
+        Helper used in partial reduction and partial sampling.
+        This marginalizes over a and returns a shifted Gaussian over b.
+        """
+        dim_a = prec_sqrt_a.shape[-2]
+        dim_b = prec_sqrt_b.shape[-2]
+        result = Tensor(
+            dim_a * math.log(2 * math.pi) / 2 - _log_det_tri(precision_chol_a),
+            int_inputs,
+        )
+        if self.rank > dim_a:
+            proj_a = _mtm(ops.triangular_solve(prec_sqrt_a, precision_chol_a))
+            prec_sqrt = prec_sqrt_b - prec_sqrt_b @ proj_a
+            white_vec = self.white_vec - _vm(self.white_vec, proj_a)
+            result += Gaussian(white_vec, prec_sqrt, inputs)
+        else:  # The Gaussian over xa is zero.
```

Comment:
Which Gaussian tests exercise this case? I want to verify that the size-0 trick used here works on all backends and doesn't introduce unexpected behavior in downstream code.

Reply:
I don't remember, maybe in one of the GaussianHMM tests?